### PR TITLE
switch Map Pair to Table 

### PR DESCRIPTION
**Pull request by loordgek**
switch Map Pair to Table,
it looks cleaner this way

### CODE REVIEW

**Comment by josephcsible** [comment]
Note that under the hood, the table is equivalent to `HashMap<Capability<?>, HashMap<EnumFacing, Object>>`. I think this will be somewhat slower than our current `HashMap<Pair<Capability<?>, EnumFacing>, Object>`.

**Comment by rubensworks** [comment]
I think so as well, we (or @loordgek) should do a quick performance comparison to see if the impact is significant.

It's quite important that it's fast, as capabilities are queried frequently.

**Comment by rubensworks** [comment]
I just hacked together the following microbenchmark:
```
		Map<Pair<Capability<?>, EnumFacing>, Object> map1 = Maps.newHashMap();
		Table<Capability<?>, EnumFacing, Object> map2 = HashBasedTable.create();
		Collection<Capability<?>> capabilities = ((IdentityHashMap<String, Capability<?>>) ReflectionHelper.getPrivateValue(CapabilityManager.class, CapabilityManager.INSTANCE, "providers")).values();
		System.out.println("caps: " + capabilities.size()); // TODO
		for (Capability<?> capability : capabilities) {
			for (EnumFacing side : EnumFacing.values()) {
				map1.put(Pair.of(capability, side), new Object());
				map2.put(capability, side, new Object());
			}
			//map1.put(Pair.of(capability, null), new Object());
			//map2.put(capability, null, new Object());
		}

		long start1 = System.currentTimeMillis();
		for (int i = 0; i < 100000; i++) {
			for (Capability<?> capability : capabilities) {
				for (EnumFacing side : EnumFacing.values()) {
					assert map1.get(Pair.of(capability, side)) != null;
				}
				//assert map1.get(Pair.of(capability, null)) != null;
				//assert map1.get(Pair.of(new Object(), null)) == null;
			}
		}
		long stop1 = System.currentTimeMillis();
		System.out.println("Duration Map: " + (stop1 - start1));

		long start2 = System.currentTimeMillis();
		for (int i = 0; i < 100000; i++) {
			for (Capability<?> capability : capabilities) {
				for (EnumFacing side : EnumFacing.values()) {
					assert map2.get(capability, side) != null;
				}
				//assert map2.get(capability, null) != null;
				//assert map2.get(new Object(), null) == null;
			}
		}
		long stop2 = System.currentTimeMillis();
		System.out.println("Duration Table: " + (stop2 - start2));
```

With the following output:
```
Duration Map: 1652
Duration Table: 1398
```

So the `Table` seems to be slightly faster.

There is however one major problem: `Table` requires non-null arguments, while we use `null` to indicate 'all sides'. So in its current form, I can't accept this PR, as it will simply crash in many cases.
One solution would be to create an additional `Map<Capability<?>, Object>` for storing these `null`-sided capabilities.

@loordgek Are you up for this?

**Review by rubensworks — CHANGES_REQUESTED** [review]
See comment above.

**Comment by loordgek** [comment]
where cant it accept null ?
i can do that

**Comment by rubensworks** [comment]
@loordgek Table crashes when null args are provided on put or get. While we provide null facings in some cases.
So on insertion and lookup the null facings must be delegated to this new map, and the non-null facings can be delegated to the Table.

**Comment by loordgek** [comment]
will do

**Comment by loordgek** [comment]
weird the get() method has the parms as @Nullable, put() in Table not but the the HashBasedTable has the annotation

**Comment by rubensworks** [comment]
Yep, it seems to be just the put that can not be null.

**Comment by loordgek** [comment]
should be good now

**Comment by rubensworks** [comment]
Thanks, will look into it tomorrow.

File: src/main/java/org/cyclops/cyclopscore/tileentity/CyclopsTileEntity.java [hunk 1]
```diff
@@ -51,7 +51,8 @@
     private boolean shouldSendUpdate = false;
     private int sendUpdateBackoff = 0;
     private final boolean ticking;
-    private Table<Capability<?>, EnumFacing, Object> capabilities = HashBasedTable.create();
+    private Map<Capability<?>, Object> capabilities = new HashMap<>();
```
Comment (rubensworks) [inline src/main/java/org/cyclops/cyclopscore/tileentity/CyclopsTileEntity.java]: Replace with `Maps.newIdentityHashMap()`, I've noticed that this is a lot faster than even an ImmutableMap. (You can then also remove the copying of this unsided map in `onLoad()`)

File: src/main/java/org/cyclops/cyclopscore/tileentity/CyclopsTileEntity.java [hunk 2]
```diff
@@ -287,18 +292,19 @@ protected EnumFacing transformFacingForRotation(EnumFacing facing) {
     }
 
     @Override
-    public boolean hasCapability(Capability<?> capability, EnumFacing facing) {
-        return (capabilities != null && capabilities.contains(capability, transformFacingForRotation(facing)))
-                || (facing != null && capabilities != null && capabilities.contains(capability, null))
+    public boolean hasCapability(@Nonnull Capability<?> capability, EnumFacing facing) {
+        return (sidedCapabilities != null && sidedCapabilities.contains(capability, transformFacingForRotation(facing)))
```
Comment (rubensworks) [inline src/main/java/org/cyclops/cyclopscore/tileentity/CyclopsTileEntity.java]: Only query `sidedCapabilities` if `facing` is not null, otherwise, query `capabilities`.

File: src/main/java/org/cyclops/cyclopscore/tileentity/CyclopsTileEntity.java [hunk 3]
```diff
@@ -287,18 +292,19 @@ protected EnumFacing transformFacingForRotation(EnumFacing facing) {
...
                 || super.hasCapability(capability, facing);
     }
 
     @Override
-    public <T> T getCapability(Capability<T> capability, EnumFacing facing) {
-        if (capabilities != null) {
-            Object value = capabilities.get(capability, transformFacingForRotation(facing));
+    @SuppressWarnings("unchecked")
+    public <T> T getCapability(@Nonnull Capability<T> capability, EnumFacing facing) {
+        if (sidedCapabilities != null) {
```
Comment (rubensworks) [inline src/main/java/org/cyclops/cyclopscore/tileentity/CyclopsTileEntity.java]: Same as above, check if `facing` is null.

File: src/main/java/org/cyclops/cyclopscore/tileentity/CyclopsTileEntity.java (hunk 1 above)
Comment (rubensworks) [reply src/main/java/org/cyclops/cyclopscore/tileentity/CyclopsTileEntity.java]: Also, can you rename this field to `innerCapabilities` (also its getter), for consistensy.

File: src/main/java/org/cyclops/cyclopscore/tileentity/CyclopsTileEntity.java [hunk 4]
```diff
@@ -327,17 +333,21 @@ public boolean hasCapability(Capability<?> capability, EnumFacing facing) {
...
 
-    protected Table<Capability<?>, EnumFacing, Object> getCapabilities() {
+    protected Map<Capability<?>, Object> getCapabilities() {
         return capabilities;
     }
 
     /**
      * Apply this interface on any tile classes that should tick.
      */
-    public static interface ITickingTile extends ITickable {
```
Comment (rubensworks) [inline src/main/java/org/cyclops/cyclopscore/tileentity/CyclopsTileEntity.java]: No need to remove this keyword.

File: src/main/java/org/cyclops/cyclopscore/tileentity/CyclopsTileEntity.java [hunk 5]
```diff
@@ -233,6 +237,7 @@ public NBTTagCompound getNBTTagCompound() {
     }
 
     @Override
+    @Nonnull
```
Comment (rubensworks) [inline src/main/java/org/cyclops/cyclopscore/tileentity/CyclopsTileEntity.java]: This does not belong in this PR.

**Review by rubensworks — CHANGES_REQUESTED** [review]

File: src/main/java/org/cyclops/cyclopscore/tileentity/CyclopsTileEntity.java (hunk 4 above)
Comment (loordgek) [reply src/main/java/org/cyclops/cyclopscore/tileentity/CyclopsTileEntity.java]: idea complains about it

File: src/main/java/org/cyclops/cyclopscore/tileentity/CyclopsTileEntity.java [hunk 6]
```diff
@@ -51,7 +48,8 @@
     private boolean shouldSendUpdate = false;
     private int sendUpdateBackoff = 0;
     private final boolean ticking;
-    private Map<Capability<?>, Object> capabilities = new HashMap<>();
+    @SuppressWarnings("unchecked")
```
Comment (rubensworks) [inline src/main/java/org/cyclops/cyclopscore/tileentity/CyclopsTileEntity.java]: Why don't you just make it checked then?

File: .gitignore [hunk 7]
```diff
@@ -25,3 +25,4 @@ keystore.jks
 gradle.properties
 build_number.properties
 changelog.txt
+/.idea/
```
Comment (rubensworks) [inline .gitignore]: No

File: src/main/java/org/cyclops/cyclopscore/tileentity/CyclopsTileEntity.java [hunk 8]
```diff
@@ -293,23 +287,25 @@ protected EnumFacing transformFacingForRotation(EnumFacing facing) {
 
     @Override
     public boolean hasCapability(@Nonnull Capability<?> capability, EnumFacing facing) {
-        return (sidedCapabilities != null && sidedCapabilities.contains(capability, transformFacingForRotation(facing)))
-                || (capabilities != null && capabilities.containsKey(capability))
-                || super.hasCapability(capability, facing);
+        if (facing != null){
```
Comment (rubensworks) [inline src/main/java/org/cyclops/cyclopscore/tileentity/CyclopsTileEntity.java]: You can inline all of this.

File: src/main/java/org/cyclops/cyclopscore/tileentity/CyclopsTileEntity.java [hunk 9]
```diff
@@ -293,23 +287,25 @@ protected EnumFacing transformFacingForRotation(EnumFacing facing) {
...
-                value = capabilities.get(capability);
-            }
-            if (value != null) {
-                return (T) value;
-            }
+        Object value;
+        if (facing != null){
+            value = sidedCapabilities.get(capability, facing);
         }
+        else value = innerCapabilities.get(capability);
```
Comment (rubensworks) [inline src/main/java/org/cyclops/cyclopscore/tileentity/CyclopsTileEntity.java]: Remove `else`, because this is now inconsistent with `hasCapability`.

**Review by rubensworks — CHANGES_REQUESTED** [review]
Not happy yet.

File: src/main/java/org/cyclops/cyclopscore/tileentity/CyclopsTileEntity.java (hunk 9 above)
Comment (loordgek) [reply src/main/java/org/cyclops/cyclopscore/tileentity/CyclopsTileEntity.java]: i need the else there

Comment (rubensworks) [reply src/main/java/org/cyclops/cyclopscore/tileentity/CyclopsTileEntity.java]: No. If value is still null at this point, you should query innerCapabilities.

File: src/main/java/org/cyclops/cyclopscore/tileentity/CyclopsTileEntity.java (hunk 8 above)
Comment (loordgek) [reply src/main/java/org/cyclops/cyclopscore/tileentity/CyclopsTileEntity.java]: how?

File: .gitignore (hunk 7 above)
Comment (loordgek) [reply .gitignore]: srry keeps sneaking in :)

File: src/main/java/org/cyclops/cyclopscore/tileentity/CyclopsTileEntity.java (hunk 8 above)
Comment (rubensworks) [reply src/main/java/org/cyclops/cyclopscore/tileentity/CyclopsTileEntity.java]: Sorry, but I don't have the time to hold your hand like this, I'm not here to help you code.

Comment (loordgek) [reply src/main/java/org/cyclops/cyclopscore/tileentity/CyclopsTileEntity.java]: return facing != null ? sidedCapabilities.contains(capability, facing) : innerCapabilities.containsKey(capability) || super.hasCapability(capability, facing);

**Comment by loordgek** [comment]
and now ??

**Comment by rubensworks** [comment]
Sorry, but I don't the time to keep reviewing this over and over again.
Again, I see major problems that you should be able to see yourself.

I'm closing this, thanks for your effort in any case.

**Comment by loordgek** [comment]
https://github.com/loordgek/CyclopsCore/commit/2ed616b79a3005311dc770e94b59fa98849b6730
last time i try
